File: financial_product_model.py
from PyQt6.QtCore import QAbstractTableModel, Qt, QModelIndex, pyqtSignal
from financial_product import FinancialProduct
from datetime import datetime  # Add datetime import

class FinancialProductModel(QAbstractTableModel):
    status_message = pyqtSignal(str)  # Renamed signal

    def __init__(self, product: FinancialProduct, parent=None):
        super().__init__(parent)
        self.product = product
        self.attributes = self.product.attributes
        self.keys = list(self.attributes.keys())

    def rowCount(self, parent=QModelIndex()):
        return len(self.keys)

    def columnCount(self, parent=QModelIndex()):
        # Now four columns: Attribute, Value, Type, Default
        return 4

    def data(self, index, role=Qt.ItemDataRole.DisplayRole):
        if not index.isValid():
            return None

        key = self.keys[index.row()]
        if role in (Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole):
            col = index.column()
            attr_data = self.attributes[key]
            
            if col == 0:  # Attribute name
                return key
            elif col == 1:  # Current value
                return str(attr_data[0])
            elif col == 2:  # Value type
                return attr_data[1].__name__
            elif col == 3:  # Default value
                return str(attr_data[2])
        return None

    def headerData(self, section, orientation, role=Qt.ItemDataRole.DisplayRole):
        if role != Qt.ItemDataRole.DisplayRole:
            return None

        if orientation == Qt.Orientation.Horizontal:
            headers = ["Attribute", "Value", "Type", "Default"]
            return headers[section] if section < len(headers) else None
        return None

    def flags(self, index):
        flags = super().flags(index)
        if index.column() == 1:
            # Make Total field non-editable
            key = self.keys[index.row()]
            if key == "Total":
                return flags & ~Qt.ItemFlag.ItemIsEditable
            return flags | Qt.ItemFlag.ItemIsEditable
        return flags & ~Qt.ItemFlag.ItemIsEditable

    def _get_timestamp(self):
        """Helper method to format current timestamp"""
        return f"[{datetime.now().strftime('%H:%M:%S')}]"

    def setData(self, index, value, role=Qt.ItemDataRole.EditRole):
        if index.column() != 1 or not index.isValid() or role != Qt.ItemDataRole.EditRole:
            return False

        key = self.keys[index.row()]
        _, expected_type, default = self.attributes[key]

        try:
            converted_value = expected_type(value)
            setattr(self.product, key.lower(), converted_value)
            self.attributes = self.product.attributes
            
            # Emit data change for current field
            self.dataChanged.emit(index, index, [role])
            
            # If Price or Quantity changed, also update Total
            if key in ("Price", "Quantity"):
                total_row = self.keys.index("Total")
                total_index = self.index(total_row, 1)
                self.dataChanged.emit(total_index, total_index, [role])

            self.status_message.emit(
                f"{self._get_timestamp()} ✓ Successfully updated {key} to {converted_value}"
            )
            return True
        except (ValueError, TypeError) as e:
            self.status_message.emit(
                f"{self._get_timestamp()} ✗ Validation error: {str(e)}"
            )
            return False
        except Exception as e:
            self.status_message.emit(
                f"{self._get_timestamp()} ✗ Unexpected error: {str(e)}"
            )
            return False

    def refresh_model(self):
        self.attributes = self.product.attributes
        self.layoutChanged.emit()

File: FinancialProductModel.cpp
#include "FinancialProductModel.h"
#include <QDateTime>
#include <QMetaType>
#include <QDebug>

FinancialProductModel::FinancialProductModel(FinancialProduct* product, QObject* parent)
    : QAbstractTableModel(parent), m_product(product)
{
    auto productAttrs = m_product->attributes();
    for(auto it = productAttrs.constBegin(); it != productAttrs.constEnd(); ++it) {
        const QVariantList& vals = it.value();
        m_attributes.insert(it.key(), {
            vals[0],
            vals[1].toInt(),
            vals[2],
            vals[3].value<std::function<void(QVariant)>>()
        });
    }
    m_keys = m_attributes.keys();
}

int FinancialProductModel::rowCount(const QModelIndex&) const
{
    return m_keys.size();
}

int FinancialProductModel::columnCount(const QModelIndex&) const
{
    return 4;
}

QVariant FinancialProductModel::data(const QModelIndex& index, int role) const
{
    if (!index.isValid())
        return QVariant();

    const QString& key = m_keys[index.row()];
    const AttributeData& attr = m_attributes[key];

    if (role == Qt::DisplayRole || role == Qt::EditRole) {
        switch (index.column()) {
            case 0: return key;
            case 1: return attr.value;
            case 2: return QMetaType(attr.type).name();
            case 3: return attr.defaultValue;
        }
    }
    return QVariant();
}

QVariant FinancialProductModel::headerData(int section, Qt::Orientation orientation, int role) const
{
    if (role != Qt::DisplayRole)
        return QVariant();

    if (orientation == Qt::Horizontal) {
        static const QStringList headers = {"Attribute", "Value", "Type", "Default"};
        return section < headers.size() ? headers[section] : QVariant();
    }
    return QVariant();
}

Qt::ItemFlags FinancialProductModel::flags(const QModelIndex& index) const
{
    auto flags = QAbstractTableModel::flags(index);
    if (index.column() == 1) {
        const QString key = m_keys[index.row()];
        if (key == "Total")
            return flags & ~Qt::ItemIsEditable;
        return flags | Qt::ItemIsEditable;
    }
    return flags & ~Qt::ItemIsEditable;
}

bool FinancialProductModel::setData(const QModelIndex& index, const QVariant& value, int role)
{
    if (index.column() != 1 || !index.isValid() || role != Qt::EditRole)
        return false;

    const QString key = m_keys[index.row()];
    AttributeData& attr = m_attributes[key];
    
    if (!attr.setter)
        return false;

    try {
        attr.setter(value);
        
        // Replace direct assignment with conversion logic
        auto productAttrs = m_product->attributes();
        m_attributes.clear();
        for(auto it = productAttrs.constBegin(); it != productAttrs.constEnd(); ++it) {
            const QVariantList& vals = it.value();
            m_attributes.insert(it.key(), {
                vals[0],
                vals[1].toInt(),
                vals[2],
                vals[3].value<std::function<void(QVariant)>>()
            });
        }
        
        emit dataChanged(index, index, {role});
        
        // Update dependent fields
        for (const auto& dependent : m_product->attributeDependencies().value(key)) {
            int depRow = m_keys.indexOf(dependent);
            if (depRow != -1) {
                QModelIndex depIndex = this->index(depRow, 1);
                emit dataChanged(depIndex, depIndex, {role});
            }
        }

        emit statusMessage(QString("[%1] ✓ Successfully updated %2 to %3")
                           .arg(getTimestamp(), key, value.toString()));
        return true;
    } catch (const std::exception& e) {
        emit statusMessage(QString("[%1] ✗ Error: %2")
                           .arg(getTimestamp(), e.what()));
        return false;
    }
}

void FinancialProductModel::refreshModel()
{
    beginResetModel();
    auto productAttrs = m_product->attributes();
    m_attributes.clear();
    
    // Convert QVariantList entries to AttributeData structs
    for(auto it = productAttrs.constBegin(); it != productAttrs.constEnd(); ++it) {
        const QVariantList& vals = it.value();
        m_attributes.insert(it.key(), {
            vals[0],  // Current value
            vals[1].toInt(),  // QMetaType ID as int
            vals[2],  // Default value
            vals[3].value<std::function<void(QVariant)>>()  // Setter
        });
    }
    
    m_keys = m_attributes.keys();
    endResetModel();
}

QString FinancialProductModel::getTimestamp() const
{
    return QDateTime::currentDateTime().toString("HH:mm:ss");
} 